client.go

```diff
@@ -77,29 +81,37 @@ func (c *client) Start() error {
 		return err
 	}
 
-	c.loop = newLoop(c, c.conn, protocol)
-	c.closed = make(chan struct{}, 1)
 	started := make(chan struct{}, 1)
 
 	go func(c *client, started chan struct{}) {
-		c.loop.Run(started)
+		c.loopMx.Lock()
+		ctx, cancel := onecontext.Merge(c.ctx, c.conn.Context())
+		c.loopCtx = &clientContext{Context: ctx}
+		c.cancelLoop = cancel
+		c.loop = newLoop(c, c.conn, protocol)
+		c.loopMx.Unlock()
+
+		c.loopCtx.SetErr(c.loop.Run(started))
+
 		c.loopMx.Lock()
 		c.loopEnded = true
+		c.cancelLoop()
 		c.loopMx.Unlock()
-		close(c.closed)
 	}(c, started)
+
 	<-started
+
 	return nil
 }
 
 func (c *client) Stop() error {
 	err := c.loop.hubConn.Close("", false)
-	c.cancel()
+	c.cancelLoop()
 	return err
 }
 
-func (c *client) Closed() <-chan struct{} {
-	return c.closed
+func (c *client) Context() context.Context {
```

Comment:
Does that really work? What happens if you have a background context and the client disconnects? Is the context wrapped into a cancelable context? I find this confusing!

Reply:
In other words: I've always regarded a context as something that the caller imposes, not as something that the callee uses to signal state.

Reply:
This context is canceled when the context passed in NewClient is canceled, when the context of the connection is canceled or when the loop ends.
See https://github.com/philippseith/signalr/blob/9d9be00d8d340a18be165db7ed817ce8678abbe0/client.go#L91
and https://github.com/philippseith/signalr/blob/6ddd5182f48073f18bccea65a03db5b8736acb2f/hubconnection.go#L35

Reply:
I understand how that works. Using the context as api does not offer any advantage over an error or struct channel. It feels very strange though and relies on https://github.com/teivah/onecontext which is not wide-spread.

Reply:
It has advantages: With the error channel, you can get the error which caused the loop to end only one time. After reading the error from the channel, it is only the state of the loop you get. With the the struct channel, you can get only the state. 
With  a context, you can get the error and the state as often as you want.
I have been using this pattern and onecontext in other (closed source) projects and not worked well.

Reply:
> In other words: I've always regarded a context as something that the caller imposes, not as something that the callee uses to signal state.

Yes, a context signals that something should be stopped. But if you have a parent context, it also signals that the parent has been stopped (and then something where the child context is used should be stopped)
You can see client.Context as something like a parent context.